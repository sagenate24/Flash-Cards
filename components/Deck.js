import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Text, View, StyleSheet, TouchableOpacity, ScrollView } from 'react-native';
import { black, gray, white, softPurp, lightBlue } from '../utils/colors';
import { MaterialCommunityIcons, EvilIcons } from '@expo/vector-icons';

// TODO: Get decks from AsnyStorage

class Deck extends Component {
  render() {
    const { title, questions } = this.props.deck;
    return (
      <ScrollView>
        <View style={styles.container}>
          <Text style={styles.cardCount}>
            {questions.length > 1
              ? questions.length + ` Cards`
              : questions.length + ` Card`}
          </Text>
          <Text style={styles.title}>{title}</Text>
          <View style={styles.optionsContainer}>
            <TouchableOpacity style={styles.metric} onPress={() => this.props.navigation.navigate(
              'Quiz',
              { currentDeck: this.props.deck }
            )}>
              <Text style={styles.header}>
                <MaterialCommunityIcons
                  name='pencil-box-outline'
                  size={34}
                  style={{ color: '#262673' }} />
              </Text>
              <Text style={[styles.subHeader, { color: lightBlue }]}>
                QUIZ
            </Text>
            </TouchableOpacity>
            <TouchableOpacity style={styles.metric} onPress={() => this.props.navigation.navigate(
              'NewCard',
              { currentDeck: this.props.deck }
            )}>
              <Text style={styles.header}>
                <MaterialCommunityIcons
                  name='cards-outline'
                  size={34}
                  style={{ color: '#262673' }} />
              </Text>
              <Text style={[styles.subHeader, { color: lightBlue }]}>
                NEW CARD
            </Text>
            </TouchableOpacity>
          </View>
          <Text style={styles.title}>Cards</Text>
          {questions.map((item) => {
            return (
              <View style={styles.item} key={item.question}>
                <Text>{item.question}</Text>
              </View>
            )
          })}
        </View>
      </ScrollView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 26,
  },
  title: {
    fontSize: 20,
    color: black,
    fontWeight: 'bold',
  },
  header: {
    textAlign: 'center',
    paddingTop: 10,
    paddingBottom: 10,
  },
  cardCount: {
    fontSize: 14,
    color: '#737373',
  },
  optionsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    marginTop: 30,
  },
  metric: {
    flex: 1,
    marginTop: 10,
    marginBottom: 20,
    marginLeft: 10,
    marginRight: 10,
    borderBottomWidth: 4,
    borderColor: lightBlue,
    backgroundColor: white,
    shadowRadius: 3,
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  subHeader: {
    fontSize: 14,
    textAlign: 'center',
    marginBottom: 5,
  },
  item: {
    backgroundColor: white,
    borderRadius: 2,
    padding: 20,
    marginLeft: 10,
    marginRight: 10,
    marginTop: 17,
    justifyContent: 'center',
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
});

function mapStateToProps(Decks, { navigation }) {
  const { entryId } = navigation.state.params;

  return {
    entryId,
    deck: Decks[entryId]
  }
}

export default connect(mapStateToProps)(Deck)