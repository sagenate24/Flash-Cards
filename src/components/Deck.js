import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Text,
  View,
  StyleSheet,
  TouchableOpacity,
  ScrollView,
} from 'react-native';
import { removeDeck } from '../utils/api';
import { addDeck } from '../actions/decks';
import { black, white, lightBlue, red, queenBlue } from '../utils/colors';
import { profanityDetector } from '../utils/helpers';

import DeckOption from './DeckOption';

class Deck extends Component {
  handleRemoveDeck = () => {
    const { remove, goBack, deck } = this.props;

    goBack();
    remove();
    removeDeck(deck.title);
  }

  generateKeyID = () => Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);

  cardLength = (questions) => {
    if (questions && questions.length) {
      if (questions.length > 1) {
        return `${questions.length} Cards  |`;
      }
      return `${questions.length} Card  |`;
    }
    return '0 Cards  |';
  }

  render() {
    const { deck, navigation, parentalControl } = this.props;

    if (deck === null) {
      return null;
    }

    const { title, questions } = deck;

    return (
      <ScrollView>
        {deck !== null
          ? (
            <View style={styles.container}>
              <View style={styles.deckTitleAndDelete}>
                <Text style={styles.cardCount}>
                  {this.cardLength(questions)}
                </Text>
                <TouchableOpacity onPress={this.handleRemoveDeck}>
                  <Text style={[styles.removeText, { marginRight: 10 }]}>DELETE</Text>
                </TouchableOpacity>
              </View>
              <Text style={styles.title}>
                {parentalControl === 'on' ? profanityDetector(title) : title}
              </Text>
              <View style={styles.optionsContainer}>
                <TouchableOpacity
                  disabled={questions.length === 0}
                  style={[styles.navOption, questions.length === 0 && { opacity: 0.6 }]}
                  onPress={() => navigation.navigate(
                    'Quiz',
                    {
                      currentDeck: deck,
                      parentalControl,
                    },
                  )}
                >
                  <DeckOption
                    size={34}
                    name="pencil-box-outline"
                    iconStyle={{ color: queenBlue }}
                    subHeaderColor={{ color: lightBlue }}
                  >
                    TAKE QUIZ
                  </DeckOption>
                </TouchableOpacity>
                <TouchableOpacity
                  style={styles.navOption}
                  onPress={() => navigation.navigate(
                    'NewCard',
                    {
                      currentDeck: deck,
                      parentalControl,
                    },
                  )}
                >
                  <DeckOption
                    size={34}
                    name="cards-outline"
                    iconStyle={{ color: queenBlue }}
                    subHeaderColor={{ color: lightBlue }}
                  >
                    ADD CARD
                  </DeckOption>
                </TouchableOpacity>
              </View>
              <Text style={styles.title}>Cards</Text>
              {questions && questions.length
                ? questions.map(item => (
                  <View style={styles.cards} key={this.generateKeyID()}>
                    <Text>
                      {parentalControl === 'on' ? profanityDetector(item.question) : item.question}
                    </Text>
                  </View>
                ))
                : <Text style={styles.empty}>No Cards</Text>}
            </View>
          )
          : null}
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 26,
  },
  deckTitleAndDelete: {
    flex: 1,
    flexDirection: 'row',
  },
  title: {
    fontSize: 20,
    color: black,
    fontWeight: 'bold',
  },
  removeText: {
    fontSize: 14,
    color: red,
    opacity: 0.8,
    marginLeft: 7,
    marginTop: 0.5,
  },
  cardCount: {
    fontSize: 14,
    color: '#737373',
  },
  optionsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    marginTop: 30,
  },
  navOption: {
    flex: 1,
    marginTop: 10,
    marginBottom: 20,
    marginLeft: 10,
    marginRight: 10,
    borderBottomWidth: 4,
    borderColor: lightBlue,
    backgroundColor: white,
    shadowRadius: 3,
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  cards: {
    backgroundColor: white,
    borderRadius: 2,
    padding: 20,
    marginLeft: 10,
    marginRight: 10,
    marginTop: 17,
    justifyContent: 'center',
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  empty: {
    color: '#C0C0C0',
    flex: 1,
    fontWeight: 'bold',
    fontSize: 19,
    alignSelf: 'center',
    marginTop: 100,
  },
});

function mapStateToProps(state, { navigation }) {
  const { currentDeck, parentalControl } = navigation.state.params;

  return {
    parentalControl,
    deck: state.decks[currentDeck.title],
  };
}

function mapDispatchToProps(dispatch, { navigation }) {
  const { currentDeck } = navigation.state.params;

  return {
    remove: () => dispatch(addDeck({
      [currentDeck.title]: null,
    })),
    goBack: () => navigation.navigate(
      'Home',
    ),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Deck);
