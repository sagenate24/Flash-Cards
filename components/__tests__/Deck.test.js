import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Deck from '../Deck';

const mockCurrentDeck = {
  questions: [{ anser: 'a', question: 'b' }],
  title: 'reactAndThings',
};

test('Deck snapShot', () => {
  const snap = renderer.create(
    <Deck.WrappedComponent deck={mockCurrentDeck} />,
  ).toJSON();

  expect(snap).toMatchSnapshot();
});

it('should flip card', () => {
  const wrapper = renderer.create(<Deck.WrappedComponent deck={mockCurrentDeck} />).getInstance();
  const cardLength = wrapper.cardLength(mockCurrentDeck.questions);

  expect(cardLength).toEqual('1 Card  |');
});
