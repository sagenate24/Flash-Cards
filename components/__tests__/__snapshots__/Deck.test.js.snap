// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match Snapshot 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "flex": 1,
          "padding": 26,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#737373",
              "fontSize": 14,
            }
          }
        >
          1 Card  |
        </Text>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#e63946",
                  "fontSize": 14,
                  "marginLeft": 7,
                  "marginTop": 0.5,
                  "opacity": 0.8,
                },
                Object {
                  "marginRight": 10,
                },
              ]
            }
          >
            DELETE
          </Text>
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#404040",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        reactAndThings
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-around",
            "marginTop": 30,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#fff",
              "borderBottomWidth": 4,
              "borderColor": "#a8dadc",
              "flex": 1,
              "marginBottom": 20,
              "marginLeft": 10,
              "marginRight": 10,
              "marginTop": 10,
              "opacity": 1,
              "shadowColor": "rgba(0,0,0,0.24)",
              "shadowOffset": Object {
                "height": 3,
                "width": 0,
              },
              "shadowOpacity": 0.8,
              "shadowRadius": 3,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "paddingBottom": 10,
                  "paddingTop": 10,
                  "textAlign": "center",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              />
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontSize": 14,
                    "marginBottom": 5,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#a8dadc",
                  },
                ]
              }
            >
              TAKE QUIZ
            </Text>
          </View>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#fff",
              "borderBottomWidth": 4,
              "borderColor": "#a8dadc",
              "flex": 1,
              "marginBottom": 20,
              "marginLeft": 10,
              "marginRight": 10,
              "marginTop": 10,
              "opacity": 1,
              "shadowColor": "rgba(0,0,0,0.24)",
              "shadowOffset": Object {
                "height": 3,
                "width": 0,
              },
              "shadowOpacity": 0.8,
              "shadowRadius": 3,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "paddingBottom": 10,
                  "paddingTop": 10,
                  "textAlign": "center",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              />
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontSize": 14,
                    "marginBottom": 5,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#a8dadc",
                  },
                ]
              }
            >
              ADD CARD
            </Text>
          </View>
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#404040",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Cards
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 2,
            "justifyContent": "center",
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 17,
            "padding": 20,
            "shadowColor": "rgba(0,0,0,0.24)",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.8,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          b
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
